File: examples/java/src/a/Demo04.java
package a;

import java.util.ArrayList;
import java.util.HashMap;

public class Demo04 {
  public static final HashMap<Character, Integer> getDigitsMap() {
    HashMap<Character, Integer> tmp1 = new HashMap<Character, Integer>();
    tmp1.put('0', 0);
    tmp1.put('1', 1);
    tmp1.put('2', 2);
    tmp1.put('3', 3);
    tmp1.put('4', 4);
    tmp1.put('5', 5);
    tmp1.put('6', 6);
    tmp1.put('7', 7);
    tmp1.put('8', 8);
    tmp1.put('9', 9);
    tmp1.put('٠', 0);
    tmp1.put('١', 1);
    tmp1.put('٢', 2);
    tmp1.put('٣', 3);
    tmp1.put('٤', 4);
    tmp1.put('٥', 5);
    tmp1.put('٦', 6);
    tmp1.put('٧', 7);
    tmp1.put('٨', 8);
    tmp1.put('٩', 9);
    tmp1.put('০', 0);
    tmp1.put('১', 1);
    tmp1.put('২', 2);
    tmp1.put('৩', 3);
    tmp1.put('৪', 4);
    tmp1.put('৫', 5);
    tmp1.put('৬', 6);
    tmp1.put('৭', 7);
    tmp1.put('৮', 8);
    tmp1.put('৯', 9);
    return tmp1;
  }

  static final HashMap<Character, Integer> digitsMap = getDigitsMap();

  public static final int parse(final String s) {
    int result = 0;
    final int strLength = s.length();
    {
      int i = 0;
      while ((i < strLength)) {
        {
          final char digit = s.charAt(i);
          {
            boolean tmp2 = digitsMap.containsKey(digit);
            if (tmp2) {
              final int digitVal = digitsMap.get(digit);
              result = ((10 * result) + digitVal);
            }
          }
        }
        i = (i + 1);
      }
      return result;
    }
  }

  public static final HashMap<String, ArrayList<Character>> getNumberSystemsMap() {
    HashMap<String, ArrayList<Character>> tmp3 = new HashMap<String, ArrayList<Character>>();
    ArrayList<Character> tmp4 = new ArrayList<Character>();
    tmp4.add('٠');
    tmp4.add('١');
    tmp4.add('٢');
    tmp4.add('٣');
    tmp4.add('٤');
    tmp4.add('٥');
    tmp4.add('٦');
    tmp4.add('٧');
    tmp4.add('٨');
    tmp4.add('٩');
    tmp3.put("ARABIC", tmp4);
    ArrayList<Character> tmp5 = new ArrayList<Character>();
    tmp5.add('০');
    tmp5.add('১');
    tmp5.add('২');
    tmp5.add('৩');
    tmp5.add('৪');
    tmp5.add('৫');
    tmp5.add('৬');
    tmp5.add('৭');
    tmp5.add('৮');
    tmp5.add('৯');
    tmp3.put("BENGALI", tmp5);
    ArrayList<Character> tmp6 = new ArrayList<Character>();
    tmp6.add('0');
    tmp6.add('1');
    tmp6.add('2');
    tmp6.add('3');
    tmp6.add('4');
    tmp6.add('5');
    tmp6.add('6');
    tmp6.add('7');
    tmp6.add('8');
    tmp6.add('9');
    tmp3.put("LATIN", tmp6);
    final HashMap<String, ArrayList<Character>> m = tmp3;
    return m;
  }

  static final HashMap<String, ArrayList<Character>> numberSystemsMap = getNumberSystemsMap();

  public static final HashMap<String, Character> getGroupingSeparatorsMap() {
    HashMap<String, Character> tmp7 = new HashMap<String, Character>();
    tmp7.put("ARABIC", '٬');
    tmp7.put("BENGALI", ',');
    tmp7.put("LATIN", ',');
    return tmp7;
  }

  static final HashMap<String, Character> groupingSeparatorsMap = getGroupingSeparatorsMap();

  public static final ArrayList<Integer> getSeparatorPositions(
      final int numLength, final String groupingStrategy) {
    ArrayList<Integer> result = new ArrayList<Integer>();
    {
      boolean tmp8 = groupingStrategy.equals("NONE");
      if (tmp8) {
        return result;
      } else {
        boolean tmp9 = groupingStrategy.equals("ON_ALIGNED_3_3");
        if (tmp9) {
          int i = (numLength - 3);
          {
            while ((0 < i)) {
              result.add(i);
              i = (i - 3);
            }
            return result;
          }
        } else {
          boolean tmp10 = groupingStrategy.equals("ON_ALIGNED_3_2");
          if (tmp10) {
            int i = (numLength - 3);
            {
              while ((0 < i)) {
                result.add(i);
                i = (i - 2);
              }
              return result;
            }
          } else {
            boolean tmp11 = groupingStrategy.equals("MIN_2");
            if (tmp11) {
              boolean tmp12 = (numLength <= 4);
              if (tmp12) {
                return result;
              } else {
                int i = (numLength - 3);
                {
                  while ((0 < i)) {
                    result.add(i);
                    i = (i - 3);
                  }
                  return result;
                }
              }
            } else {
              return result;
            }
          }
        }
      }
    }
  }

  public static final String format(
      final int num, final String numberSystem, final String groupingStrategy) {
    int i = num;
    java.lang.StringBuffer result = new StringBuffer();
    {
      while (!(i == 0)) {
        final int quotient = (i / 10);
        final int remainder = (i % 10);
        final ArrayList<Character> numberSystemDigits = numberSystemsMap.get(numberSystem);
        final char localDigit = numberSystemDigits.get(remainder);
        {
          result.insert(0, localDigit);
          i = quotient;
        }
      }
      {
        final char sep = groupingSeparatorsMap.get(numberSystem);
        final int numLength = result.length();
        final ArrayList<Integer> separatorPositions =
            getSeparatorPositions(numLength, groupingStrategy);
        final int numPositions = separatorPositions.size();
        int idx = 0;
        while ((idx < numPositions)) {
          {
            final int position = separatorPositions.get(idx);
            result.insert(position, sep);
          }
          idx = (idx + 1);
        }
      }
      return result.toString();
    }
  }

  public static final void main(String[] _args) {
    System.out.println(parse("٥٠٣٠١"));
    System.out.println(parse("৫০৩০১"));
    System.out.println(parse("7,654,321"));
    System.out.println(parse("76,54,321"));
    System.out.println(format(7654321, "LATIN", "ON_ALIGNED_3_2"));
    System.out.println(format(7654321, "ARABIC", "ON_ALIGNED_3_3"));
    System.out.println(format(7654321, "BENGALI", "ON_ALIGNED_3_3"));
  }
}

File: examples/src/a/demo04.clj
(ns a.demo04
  (:refer-clojure :exclude [format]))

(def ^{:kalias {:mmap [:char :int]}} DigitsMap)
(def ^{:kalias {:mmap [:string {:mvector [:char]}]}} NumberSystemsMap)
(def ^{:kalias {:mvector [:char]}} NumberSystemDigits)
(def ^{:kalias {:mmap [:string :char]}} GroupingSeparatorsMap)
(def ^{:kalias {:mvector [:int]}} SeparatorPositions)

(defn getDigitsMap ^{:t DigitsMap} []
  ^{:t DigitsMap}
  {\0     (int 0)
   \1     (int 1)
   \2     (int 2)
   \3     (int 3)
   \4     (int 4)
   \5     (int 5)
   \6     (int 6)
   \7     (int 7)
   \8     (int 8)
   \9     (int 9)
   \u0660 (int 0)
   \u0661 (int 1)
   \u0662 (int 2)
   \u0663 (int 3)
   \u0664 (int 4)
   \u0665 (int 5)
   \u0666 (int 6)
   \u0667 (int 7)
   \u0668 (int 8)
   \u0669 (int 9)
   \u09E6 (int 0)
   \u09E7 (int 1)
   \u09E8 (int 2)
   \u09E9 (int 3)
   \u09EA (int 4)
   \u09EB (int 5)
   \u09EC (int 6)
   \u09ED (int 7)
   \u09EE (int 8)
   \u09EF (int 9)})

(def ^{:t DigitsMap} digitsMap (getDigitsMap))

(defn parse ^Integer [^String s]
  (let [result (atom (int 0))
        ^{:t :int} strLength (count s)]
    (dotimes [^{:t :int} i strLength]
      (let [^{:t :char} digit (nth s i)]
        (if (contains? digitsMap digit)
          (let [^Integer digitVal (get digitsMap digit)]
            (reset! result (+ (* (int 10) @result) digitVal))))))
    @result))

(defn getNumberSystemsMap
  ^{:t NumberSystemsMap} []
  (let [^{:t NumberSystemsMap} m
        {"LATIN"   [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9]
         "ARABIC"  [\u0660 \u0661 \u0662 \u0663 \u0664 \u0665 \u0666 \u0667 \u0668 \u0669]
         "BENGALI" [\u09E6 \u09E7 \u09E8 \u09E9 \u09EA \u09EB \u09EC \u09ED \u09EE \u09EF]}]
    m))

(def ^{:t NumberSystemsMap}
  numberSystemsMap (getNumberSystemsMap))

(defn getGroupingSeparatorsMap
  ^{:t GroupingSeparatorsMap} []
  ^{:t GroupingSeparatorsMap}
  {"LATIN"   \,
   "ARABIC"  \٬
   "BENGALI" \,})

(def ^{:t GroupingSeparatorsMap}
  groupingSeparatorsMap (getGroupingSeparatorsMap))

(defn getSeparatorPositions ^{:t SeparatorPositions}
  [^Integer numLength ^String groupingStrategy]
  (let [^{:t SeparatorPositions} result (atom [])]
    (cond
      (= groupingStrategy "NONE")
      @result

      (= groupingStrategy "ON_ALIGNED_3_3")
      (let [^{:t :int} i (atom (- numLength (int 3)))]
        (while (< (int 0) @i)
          (swap! result conj @i)
          (reset! i (- @i (int 3))))
        @result)

      (= groupingStrategy "ON_ALIGNED_3_2")
      (let [^{:t :int} i (atom (- numLength (int 3)))]
        (while (< (int 0) @i)
          (swap! result conj @i)
          (reset! i (- @i (int 2))))
        @result)

      (= groupingStrategy "MIN_2")
      (if (<= numLength (int 4))
        @result
        (let [^{:t :int} i (atom (- numLength (int 3)))]
          (while (< (int 0) @i)
            (swap! result conj @i)
            (reset! i (- @i (int 3))))
          @result))

      :else
      @result)))

(defn format
  ^String [^Integer num, ^String numberSystem, ^String groupingStrategy]
  (let [^{:t :int} i (atom num)
        ^:mut ^StringBuffer result (StringBuffer.)]
    (while (not (= @i (int 0)))
      (let [^Integer quotient (quot @i (int 10))
            ^Integer remainder (rem @i (int 10))
            ^{:t NumberSystemDigits} numberSystemDigits (get numberSystemsMap numberSystem)
            ^{:t :char} localDigit (nth numberSystemDigits remainder)]
        (.insert result (int 0) localDigit)
        (reset! i quotient)))
    (let [^{:t :char} sep (get groupingSeparatorsMap numberSystem)
          ^{:t :int} numLength (.length result)
          ^{:t SeparatorPositions} separatorPositions (getSeparatorPositions numLength groupingStrategy)
          ^{:t :int} numPositions (count separatorPositions)]
      (dotimes [^{:t :int} idx numPositions]
        (let [^int position (nth separatorPositions idx)]
          (.insert result position sep))))
    (.toString result)))

(defn -main ^{:t :void} [& _args]
  (println (parse "\u0665\u0660\u0663\u0660\u0661"))
  (println (parse "৫০৩০১"))
  (println (parse "7,654,321"))
  (println (parse "76,54,321"))

  (println (format (int 7654321) "LATIN" "ON_ALIGNED_3_2"))
  (println (format (int 7654321) "ARABIC" "ON_ALIGNED_3_3"))
  (println (format (int 7654321) "BENGALI" "ON_ALIGNED_3_3")))
